### PR TITLE
[FLINK-33576][core] Introduce new Flink conf file 'config.yaml' supporting standard YAML syntax.

### PR DESCRIPTION
**Pull request by JunRuiLee**
## What is the purpose of the change

Introduce new Flink conf file 'config.yaml' supporting standard YAML syntax.

## Brief change log

  - *Introduce new Flink conf file 'config.yaml' supporting standard YAML syntax.*
  - *Change the UT case to use new conf file.*

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java [hunk 1]
```diff
@@ -112,25 +112,26 @@ public static String[] parseLocalStateDirectories(Configuration configuration) {
     }
 
     /**
-     * Parses a string as a map of strings. The expected format of the map is:
+     * Parses a string as a map of strings. The expected format of the map which parsed by FLINK
```
Comment (zhuzhurk) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java:115]: which parsed -> to be parsed

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java [hunk 2]
```diff
@@ -112,25 +112,26 @@ public static String[] parseLocalStateDirectories(Configuration configuration) {
...
+     * parser is:
      *
      * <pre>
      * key1:value1,key2:value2
      * </pre>
      *
+     * <p>The expected format of the map which parsed by standard YAML parser is:
```
Comment (zhuzhurk) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java:122]: which parsed -> to be parsed

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/RowLevelUpdateTest.java [hunk 3]
```diff
@@ -136,7 +136,7 @@ void testUpdateWithCustomColumns() {
                                 "CREATE TABLE t (a int, b string, c double) WITH"
                                         + " ("
                                         + "'connector' = 'test-update-delete', "
-                                        + "'required-columns-for-update' = 'b;c', "
+                                        + "'required-columns-for-update' = '[b, c]', "
```
Comment (zhuzhurk) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/RowLevelUpdateTest.java:139]: The incompatibility of existing SQLs can be a problem.
I think we need to find way that allows users to use standard YAML config file without changing the existing sql jobs.

**Review by zhuzhurk — COMMENTED** [review]
Thanks for creating this PR! @JunRuiLee 

There are still some `flink-conf` presents in the code, including `YarnTestBase` and many comments. I think they should also be changed.

I also think we need to find a better way for SQL compatibility.

File: flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/FlinkDistribution.java [hunk 4]
```diff
@@ -422,7 +422,10 @@ public void appendConfiguration(Configuration config) throws IOException {
                         .map(entry -> entry.getKey() + ": " + entry.getValue())
                         .collect(Collectors.toList());
 
-        Files.write(conf.resolve("flink-conf.yaml"), configurationLines);
+        // NOTE: Before we change the default conf file in the flink-dist to 'config.yaml', we
+        // need to use the legacy flink conf file 'flink-conf.yaml' here.
```
[the remark this answers is not in the document]
Comment (JunRuiLee) [reply flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/FlinkDistribution.java:426]: Yes

**Comment by JunRuiLee** [comment]
Hi @zhuzhurk , Thanks for reviewing. I've updated the PR with the necessary changes. Please take a look~

Here are some clarifications I'd like to make:

1. Concerning the thread safety issues of the GlobalConfiguration.isStandardYaml state during concurrent tests, the current testing framework in Flink ensures tests are run sequentially between different test classes within the same process. This setup means we only need to ensure the consistency of that state upon the completion of each test class.

2. Regarding the updates needed for comments within the code and specific code references to "flink-conf.yaml" (for example, in YarnTestBase), I intend to address these adjustments in a follow-up task. This subsequent task will involve updating the default configuration file in flink-dist to config.yaml, at which time I plan to collectively revise both the comments in the code and the relevant code references.

File: flink-core/src/main/java/org/apache/flink/configuration/GlobalConfiguration.java [hunk 5]
```diff
@@ -241,6 +265,95 @@ private static Configuration loadYAMLResource(File file) {
...
+     * keyA.keyB.keyD: "world"
+     * </pre>
+     *
+     * @param config an arbitrarily nested config map
+     * @param keyPrefix The string to prefix the keys in the current config level
+     * @return A flattened, 1 level deep map
+     */
+    @SuppressWarnings("unchecked")
+    private static Map<String, Object> flatten(
+            @Nonnull Map<String, Object> config, String keyPrefix) {
```
Comment (zhuzhurk) [inline flink-core/src/main/java/org/apache/flink/configuration/GlobalConfiguration.java:293]: Flink parameters are by default non-null. Only nullable ones should be explicitly annotated with `@Nullable`.

**Review by zhuzhurk — COMMENTED** [review]
Thanks for addressing the comments. @JunRuiLee 
The change looks good to me except for one minor comment.

**Comment by JunRuiLee** [comment]
@zhuzhurk I've updated this pr accordingly, PTAL.

**Comment by JunRuiLee** [comment]
@flinkbot run azure

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Comment by JunRuiLee** [comment]
@flinkbot run azure

**Pull request closed without merging**